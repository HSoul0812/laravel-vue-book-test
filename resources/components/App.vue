<template>
	<h1 class="text-center mt-5">Laravel Vue Book List</h1>
	<div class="mx-5 d-flex align-items-end my-5">
		<div class="ms-auto">
			<b-button class="mx-1" variant="outline-primary" @click="onAddBook">Add
				Book</b-button>
			<b-button class="mx-1" variant="outline-primary" @click="onAddAuthor">Add
				Author</b-button>
			<b-button class="mx-1 " variant="outline-primary" @click="onAddLibrary">Add
				Library</b-button>
		</div>
	</div>
	<EasyDataTable :headers="headers" :items="books" class="px-5" data-test="testTable">
		<template #item-operation="item">
			<div class="operation-wrapper">
				<img src="../assets/image/delete.png" class="operation-icon" @click="deleteItem(item)" />
				<img src="../assets/image/edit.png" class="operation-icon" @click="onEdit(item)" />
			</div>
		</template>
	</EasyDataTable>

	<book-modal :show="bookModalShow" :currentItem="currentItem" :authors="authors" :libraries="libraries"
		@addItem="addBook" @updateItem="updateBook" @updateShow="(v) => bookModalShow = v"></book-modal>

	<author-modal :show="authorModalShow" @addItem="addAuthor" @updateShow="(v) => authorModalShow = v"></author-modal>

	<library-modal :show="libraryModalShow" @addItem="addLibrary"
		@updateShow="(v) => libraryModalShow = v"></library-modal>
</template>
<script>

import Service from '../service/Service';
import BookModal from './BookModal.vue';
import AuthorModal from './AuthorModal.vue';
import LibraryModal from './LibraryModal.vue';
import { formatDate } from '../utils/common';

export default {
	name: 'App',
	components: {
		BookModal,
		AuthorModal,
		LibraryModal
	},
	data: function () {
		return {
			books: [],
			authors: [],
			libraries: [],
			currentItem: {},
			bookModalShow: false,
			authorModalShow: false,
			libraryModalShow: false,
			addClicked: false,
			headers: [
				{ text: 'Book Name', value: "name" },
				{ text: 'Year', value: "year" },
				{ text: 'Author Name', value: "authorName" },
				{ text: 'Genre', value: "authorGenre" },
				{ text: 'Library Name', value: "libraryName" },
				{ text: 'Address', value: "libraryAddress" },
				{ text: "Operation", value: "operation" },
			],
		}
	},

	methods: {
		onAddBook() {
			this.bookModalShow = !this.bookModalShow;
			this.currentItem = {};
		},

		onAddAuthor() {
			this.authorModalShow = !this.authorModalShow;
			this.currentItem = {};
		},

		onAddLibrary() {
			this.libraryModalShow = !this.libraryModalShow;
			this.currentItem = {};
		},

		onEdit(item) {
			this.bookModalShow = !this.bookModalShow;
			this.currentItem = item;
		},

		async loadItems() {
			this.books = await Service.loadBooks();
			this.authors = await Service.loadAuthors();
			this.libraries = await Service.loadLibraries();
		},

		async deleteItem(item) {
			this.books = await Service.deleteBook(item.id);
		},

		async addLibrary(item) {
			this.libraries = await Service.addLibrary(item);
		},

		async addAuthor(item) {
			this.authors = await Service.addAuthor({ ...item, authorBirth: formatDate(item.authorBirth) });
		},

		async addBook(item) {
			this.books = await Service.addBook(item);
		},

		async updateBook(item) {
			this.books = await Service.updateBook(item, item.id);
		},
	},

	computed: {
		authorOptions() {
			return this.authors.map(author => ({ value: author.id, text: author.name }))
		},

		authorChange: {
			get() {
				return this.currentItem.author_id;
			},
			set(newValue) {
				const newAuthor = this.authors.find(author => author.id === newValue);

				if (newAuthor) {
					this.currentItem.author_id = newValue;
					this.currentItem.authorName = newAuthor.name;
					this.currentItem.authorGenre = newAuthor.genre;
					this.currentItem.authorBirth = newAuthor.birthday;
				}
			}
		},

		libraryOptions() {
			return this.libraries.map(library => ({ value: library.id, text: library.name }))
		},

		libraryChange: {
			get() {
				return this.currentItem.libraryId;
			},
			set(newValue) {
				const newLibrary = this.libraries.find(library => library.id === newValue);

				if (newLibrary) {
					this.currentItem.libraryId = newValue;
					this.currentItem.libraryName = newLibrary.name;
					this.currentItem.libraryAddress = newLibrary.address;
				}
			}
		},
	},

	created() {
		this.loadItems();
	}
}
</script>
