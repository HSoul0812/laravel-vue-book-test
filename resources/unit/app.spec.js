import { mount } from '@vue/test-utils'
import App from '../components/App'

describe('App Vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(App)
  })

  test('renders properly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

})
