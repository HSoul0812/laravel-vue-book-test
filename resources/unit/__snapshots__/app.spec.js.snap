// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Vue renders properly 1`] = `
"<h1 class=\\"text-center mt-5\\">Laravel Vue Book List</h1>
<div class=\\"mx-5 d-flex align-items-end my-5\\">
  <div class=\\"ms-auto\\">
    <b-button class=\\"mx-1\\" variant=\\"outline-primary\\">Add Book</b-button>
    <b-button class=\\"mx-1\\" variant=\\"outline-primary\\">Add Author</b-button>
    <b-button class=\\"mx-1\\" variant=\\"outline-primary\\">Add Library</b-button>
  </div>
</div>
<easydatatable headers=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" items=\\"\\" class=\\"px-5\\" data-test=\\"testTable\\"></easydatatable>
<b-modal modelvalue=\\"false\\">
  <h4 class=\\"text-center\\">Edit Book</h4>
  <form class=\\"px-5\\">
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Book Name : </div>
        <b-form-input class=\\"w-50 mx-auto\\" data-test=\\"bookTest\\" disabled=\\"false\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Year : </div>
        <b-form-input class=\\"w-50 mx-auto\\" data-test=\\"bookTest\\" disabled=\\"false\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Author Name : </div>
        <b-form-select class=\\"w-50\\" options=\\"\\"></b-form-select>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Genre : </div>
        <b-form-input class=\\"w-50 mx-auto\\" data-test=\\"bookTest\\" disabled=\\"true\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Library Name : </div>
        <b-form-select class=\\"w-50\\" options=\\"\\"></b-form-select>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Address : </div>
        <b-form-input class=\\"w-50 mx-auto\\" data-test=\\"bookTest\\" disabled=\\"true\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <!--v-if-->
    </div>
  </form>
</b-modal>
<b-modal modelvalue=\\"false\\">
  <h4 class=\\"text-center\\">Input Author Info</h4>
  <form class=\\"px-5\\">
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Author Name : </div>
        <b-form-input class=\\"w-50 mx-auto\\" disabled=\\"false\\" data-test=\\"authorTest\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Genre : </div>
        <b-form-input class=\\"w-50 mx-auto\\" disabled=\\"false\\" data-test=\\"authorTest\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Birthday : </div>
        <b-form-input class=\\"w-50 mx-auto\\" disabled=\\"false\\" data-test=\\"authorTest\\" required=\\"\\"></b-form-input>
      </div>
    </div>
  </form>
</b-modal>
<b-modal modelvalue=\\"false\\">
  <h4 class=\\"text-center\\">Input Library Info</h4>
  <form class=\\"px-5\\">
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Library Name : </div>
        <b-form-input class=\\"w-50 mx-auto\\" disabled=\\"false\\" data-test=\\"libraryTest\\" required=\\"\\"></b-form-input>
      </div>
    </div>
    <div>
      <div class=\\"d-flex align-items-center justify-content-between my-2\\">
        <div class=\\"w-50\\">Library Address : </div>
        <b-form-input class=\\"w-50 mx-auto\\" disabled=\\"false\\" data-test=\\"libraryTest\\" required=\\"\\"></b-form-input>
      </div>
    </div>
  </form>
</b-modal>"
`;
